File: src/lib/spotify-api.ts
// Dünner Wrapper um die Spotify Web API. Alle Calls holen sich einen frischen Token.
// Siehe docs/tech-stack.md → API-Calls die wir brauchen.
import { getValidAccessToken } from './spotify-auth';
import type { Track, TrackSource } from '../types';

const API = 'https://api.spotify.com/v1';

async function api<T>(path: string, init?: RequestInit): Promise<T> {
  const token = await getValidAccessToken();
  const res = await fetch(`${API}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
      ...(init?.headers ?? {}),
    },
  });
  if (res.status === 429) {
    const retry = Number(res.headers.get('Retry-After') ?? '1');
    await new Promise((r) => setTimeout(r, (retry + 0.5) * 1000));
    return api<T>(path, init);
  }
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Spotify API ${res.status} bei ${path}: ${text}`);
  }
  // Manche Playback-Endpoints liefern 204 ohne Body.
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

// ---- /me ----
export type SpotifyUser = {
  id: string;
  display_name: string | null;
  email?: string;
  product: string; // "premium" | "free" | ...
  images?: { url: string }[];
};

export function getMe(): Promise<SpotifyUser> {
  return api<SpotifyUser>('/me');
}

// ---- Playlists ----
export type SpotifyPlaylist = {
  id: string;
  name: string;
  images: { url: string }[];
  // Spotify hat das Track-Collection-Feld von `tracks` auf `items` umgestellt;
  // ältere Accounts liefern evtl. noch `tracks`. Beide optional behandeln.
  tracks?: { total: number };
  items?: { total: number };
  owner: { display_name: string | null; id: string };
  collaborative: boolean;
};

/** Song-Anzahl einer Playlist, egal ob Spotify `tracks` oder `items` liefert. */
export function playlistTrackTotal(p: SpotifyPlaylist): number {
  return (p.tracks ?? p.items)?.total ?? 0;
}

type Paged<T> = { items: T[]; next: string | null };

export async function getMyPlaylists(): Promise<SpotifyPlaylist[]> {
  // Map nach id: filtert null/kaputte Einträge UND Pagination-Duplikate (Spotify liefert
  // bei großen Listen dieselbe Playlist gelegentlich auf mehreren Seiten).
  const byId = new Map<string, SpotifyPlaylist>();
  let url = '/me/playlists?limit=50';
  for (;;) {
    const page = await api<Paged<SpotifyPlaylist>>(url);
    for (const p of page.items) {
      if (p && p.id && !byId.has(p.id)) byId.set(p.id, p);
    }
    if (!page.next) break;
    url = page.next.replace(API, '');
  }
  return [...byId.values()];
}

// ---- Tracks einer Playlist ----
type PlaylistTrackItem = {
  added_by?: { id: string } | null;
  // Spotify hat den Track-Wrapper umbenannt (tracks→items-Umstellung). Mehrere Keys
  // tolerieren, damit alte wie neue Antworten funktionieren.
  track?: SpotifyTrackObject | null;
  item?: SpotifyTrackObject | null;
};

function extractTrack(item: PlaylistTrackItem): SpotifyTrackObject | null {
  return item.track ?? item.item ?? null;
}

type SpotifyTrackObject = {
  id: string | null;
  uri: string;
  name: string;
  duration_ms: number;
  explicit: boolean;
  is_playable?: boolean;
  artists: { name: string }[];
  album: {
    name: string;
    images: { url: string }[];
    release_date: string;
    release_date_precision: 'year' | 'month' | 'day';
  };
};

function parseYear(releaseDate: string): number {
  const y = parseInt(releaseDate.slice(0, 4), 10);
  return Number.isFinite(y) ? y : 0;
}

function toTrack(t: SpotifyTrackObject, source: TrackSource, addedById?: string): Track {
  return {
    id: t.id ?? t.uri,
    uri: t.uri,
    name: t.name,
    artist: t.artists.map((a) => a.name).join(', '),
    albumName: t.album.name,
    albumArt: t.album.images?.[0]?.url ?? '',
    releaseYear: parseYear(t.album.release_date),
    releaseDateRaw: t.album.release_date,
    releaseDatePrecision: t.album.release_date_precision,
    source,
    addedById,
    durationMs: t.duration_ms,
    explicit: t.explicit,
  };
}

async function fetchPlaylistItems(
  playlistId: string,
  endpoint: 'items' | 'tracks',
  source: TrackSource,
): Promise<Track[]> {
  const out: Track[] = [];
  // Kein fields-Filter: Spotify hat Feldnamen umbenannt; volle Objekte holen und
  // robust extrahieren (extractTrack toleriert `track` wie `item`).
  let url = `/playlists/${playlistId}/${endpoint}?limit=100`;
  for (;;) {
    const page = await api<Paged<PlaylistTrackItem>>(url);
    for (const item of page.items) {
      const t = extractTrack(item);
      if (!t || !t.id) continue; // lokale Dateien / entfernte Tracks raus
      if (t.is_playable === false) continue;
      out.push(toTrack(t, source, item.added_by?.id));
    }
    if (!page.next) break;
    url = page.next.replace(API, '');
  }
  return out;
}

/**
 * Holt alle Tracks einer Playlist (paginiert, 100/Call) und mappt sie auf unser Track-Modell.
 * Spotify hat den Sub-Endpoint von `/tracks` auf `/items` umgestellt — wir versuchen `/items`
 * zuerst und fallen bei 403/404 auf das alte `/tracks` zurück (für ältere Accounts).
 */
export async function getPlaylistTracks(
  playlistId: string,
  source: TrackSource,
): Promise<Track[]> {
  try {
    return await fetchPlaylistItems(playlistId, 'items', source);
  } catch (e) {
    const msg = e instanceof Error ? e.message : '';
    if (/\b(403|404)\b/.test(msg)) {
      return fetchPlaylistItems(playlistId, 'tracks', source);
    }
    throw e;
  }
}

// ---- Playback (Web Playback SDK device) ----

export function transferPlayback(deviceId: string, play = false): Promise<void> {
  return api<void>('/me/player', {
    method: 'PUT',
    body: JSON.stringify({ device_ids: [deviceId], play }),
  });
}

export function playTrack(
  deviceId: string,
  uri: string,
  positionMs = 0,
): Promise<void> {
  return api<void>(`/me/player/play?device_id=${deviceId}`, {
    method: 'PUT',
    body: JSON.stringify({ uris: [uri], position_ms: positionMs }),
  });
}

export function pausePlayback(deviceId: string): Promise<void> {
  return api<void>(`/me/player/pause?device_id=${deviceId}`, { method: 'PUT' });
}

export function seek(deviceId: string, positionMs: number): Promise<void> {
  return api<void>(
    `/me/player/seek?position_ms=${positionMs}&device_id=${deviceId}`,
    { method: 'PUT' },
  );
}

File: ios/Sources/Music/SpotifyAPI.swift
// Spotify Web API: eigene Playlists und Playlist-Tracks (mit Pagination) —
// portiert aus src/lib/spotify-api.ts der Web-App.
import Foundation

struct SpotifyPlaylist: Identifiable, Hashable {
    let id: String
    let name: String
    let trackCount: Int
}

@MainActor
final class SpotifyAPI {
    private let auth: SpotifyAuth

    init(auth: SpotifyAuth) {
        self.auth = auth
    }

    func myPlaylists() async throws -> [SpotifyPlaylist] {
        var out: [SpotifyPlaylist] = []
        var url: URL? = URL(string: "https://api.spotify.com/v1/me/playlists?limit=50")
        while let pageURL = url {
            let page: Page<PlaylistItem> = try await get(pageURL)
            out += page.items.map {
                SpotifyPlaylist(id: $0.id, name: $0.name, trackCount: $0.count)
            }
            url = page.next
        }
        return out
    }

    /// Spotify hat den alten Sub-Endpoint `/tracks` für viele Apps gesperrt (403);
    /// `/items` ist der aktuelle und liefert dieselben Daten. Erst `/items`, bei
    /// 403/404 Fallback auf `/tracks` (genau wie die Web-App, src/lib/spotify-api.ts).
    func playlistTracks(_ playlistId: String) async throws -> [Track] {
        do {
            return try await fetchItems(playlistId, endpoint: "items")
        } catch SpotifyError.forbidden {
            return try await fetchItems(playlistId, endpoint: "tracks")
        } catch SpotifyError.http(404) {
            return try await fetchItems(playlistId, endpoint: "tracks")
        }
    }

    private func fetchItems(_ playlistId: String, endpoint: String) async throws -> [Track] {
        var out: [Track] = []
        var url: URL? = URL(string: "https://api.spotify.com/v1/playlists/\(playlistId)/\(endpoint)?limit=100")
        while let pageURL = url {
            let page: Page<PlaylistTrackItem> = try await get(pageURL)
            for item in page.items {
                // /items legt das Track-Objekt unter `item`, /tracks unter `track`.
                guard let t = item.track ?? item.item, let id = t.id,
                      item.isLocal != true, t.isPlayable != false else { continue }
                let year = Int(String((t.album?.releaseDate ?? "").prefix(4))) ?? 0
                out.append(Track(
                    id: id,
                    uri: t.uri ?? "",
                    name: t.name,
                    artist: t.artists.map(\.name).joined(separator: ", "),
                    albumName: t.album?.name ?? "",
                    albumArt: t.album?.images?.first?.url ?? "",
                    releaseYear: year,
                    durationMs: t.durationMs,
                    explicit: t.explicit ?? false,
                    source: .friends,
                    addedByName: item.addedBy?.id
                ))
            }
            url = page.next
        }
        return out
    }

    // MARK: HTTP + Decoding

    private func get<T: Decodable>(_ url: URL) async throws -> T {
        var req = URLRequest(url: url)
        let token = try await auth.validAccessToken()
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, response) = try await URLSession.shared.data(for: req)
        let status = (response as? HTTPURLResponse)?.statusCode ?? -1
        guard status == 200 else {
            // Spotify gibt für App-Modus-Sperren ein nacktes 403 „Forbidden" zurück —
            // betrifft das Auslesen von Playlist-Inhalten (siehe SpotifyProvider).
            if status == 403 { throw SpotifyError.forbidden }
            throw SpotifyError.http(status)
        }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try decoder.decode(T.self, from: data)
    }

    private struct Page<Item: Decodable>: Decodable {
        let items: [Item]
        let next: URL?
    }

    private struct PlaylistItem: Decodable {
        struct CountRef: Decodable { let total: Int? }
        struct Owner: Decodable { let id: String? }
        let id: String
        let name: String
        let tracks: CountRef?
        let items: CountRef?  // Spotify liefert die Anzahl mittlerweile hier
        let owner: Owner?
        var count: Int { tracks?.total ?? items?.total ?? 0 }
    }

    private struct PlaylistTrackItem: Decodable {
        struct AddedBy: Decodable { let id: String? }
        let track: TrackObject?   // /tracks-Endpoint
        let item: TrackObject?    // /items-Endpoint
        let isLocal: Bool?        // liegt beim /items-Endpoint außen am Element
        let addedBy: AddedBy?
    }

    private struct TrackObject: Decodable {
        struct Artist: Decodable { let name: String }
        struct Image: Decodable { let url: String }
        struct Album: Decodable {
            let name: String?
            let releaseDate: String?
            let images: [Image]?
        }
        let id: String?
        let uri: String?
        let name: String
        let durationMs: Int
        let explicit: Bool?
        let isLocal: Bool?
        let isPlayable: Bool?
        let artists: [Artist]
        let album: Album?
    }
}
